src/device/array.jl

```diff
@@ -0,0 +1,155 @@
+# Contiguous on-device arrays
+
+export VeDeviceArray, VeDeviceVector, VeDeviceMatrix, ldg
+
+
+## construction
+
+"""
+    VeDeviceArray(dims, ptr)
+    VeDeviceArray{T}(dims, ptr)
+    VeDeviceArray{T,N}(dims, ptr)
+    VeDeviceArray{T,N,A}(dims, ptr)
+
+Construct an `N`-dimensional dense VEDA device array with element type `T` wrapping a
+pointer, where `N` is determined from the length of `dims` and `T` is determined from the
+type of `ptr`. `dims` may be a single scalar, or a tuple of integers corresponding to the
+lengths in each dimension). If the rank `N` is supplied explicitly as in `Array{T,N}(dims)`,
+then it must match the length of `dims`. The same applies to the element type `T`, which
+should match the type of the pointer `ptr`.
+"""
+VeDeviceArray
+
+# NOTE: we can't support the typical `tuple or series of integer` style construction,
+#       because we're currently requiring a trailing pointer argument.
+
+struct VeDeviceArray{T,N,A} <: AbstractArray{T,N}
```

Comment:
@efocht does the VE have different address-spaces to worry about?

Reply:
That depends. Currently you start a context on one VE. That lives in a process (/opt/nec/ve/bin/ps -ef). Once you start using multiple procs or multiple VEs, there will be multiple address spaces.

Reply:
I think Valentin is referring to LLVM addrspaces, which are important when implementing accessors for `VeDevicePtr` and `VeDeviceArray` because LLVM may need to emit different hardware instructions (this is the case on the NVPTX and AMDGPU targets).

Reply:
Anyway, I'm going to drop this file because it's not needed for basic memory management, and I wouldn't want to commit it without a bunch of test coverage.